### PR TITLE
Merge train and analyze configs

### PR DESCRIPTION
**Pull request by zurk**
Closes https://github.com/src-d/lookout-sdk-ml/issues/17 on style analyzer side. 

I also move our 70 line config to the separate file. 

A lot of files changed, but everything is just about adopting the code to new config structure.

### CODE REVIEW

**Comment by zurk** [comment]
Long-running tests in my fork: https://travis-ci.org/zurk/style-analyzer/builds/498855555?

File: lookout/style/format/analyzer.py [hunk 1]
```diff
@@ -136,8 +77,8 @@ def __init__(self, model: FormatModel, url: str, config: Mapping[str, Any]) -> N
         """
         super().__init__(model, url, config)
         self._log = logging.getLogger(type(self).__name__)
-        self.config = self._load_analyze_config(self.config)
-        with open(self.config["comment_template"], encoding="utf-8") as f:
+        self.config = self._load_config(config)
+        with open(self.config["common"]["comment_template"], encoding="utf-8") as f:
```
Comment (vmarkovtsev) [inline lookout/style/format/analyzer.py]: Why do we need the comment template in "common"?

File: lookout/style/format/benchmarks/evaluate_smoke.py [hunk 2]
```diff
@@ -328,14 +327,11 @@ def evaluate_smoke_entry(inputpath: str, reportdir: str, database: str, bblfsh:
                     repopath = inputpath / row["repo"]
                     config_json = {
                         SmokeEvalFormatAnalyzer.name:
-                            merge_dicts(analyze_config, {
+                            merge_dicts(config, {"common": {
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/evaluate_smoke.py]: IDK the specifics here, but the things like repository name should be excluded from all the configuration. The only source of truth is what is embedded into the model file.

File: lookout/style/format/benchmarks/general_report.py [hunk 3]
```diff
@@ -257,13 +256,13 @@ def analyze(self, ptr_from: ReferencePointer, ptr_to: ReferencePointer,
             filepath = fix.head_file.path
             if fix.error:
                 continue
-            if self.config["aggregate"]:
+            if self.config["common"]["aggregate"]:
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/general_report.py]: Why common?

File: lookout/style/format/default_config.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
+import os
+
+DEFAULT_STYLE_ANALYZER_CONFIG = {
+    "common": {
+        # All common for train and analyze config values for any language should be here
+        "comment_template": os.path.join(os.path.dirname(__file__), "templates", "comment.jinja2"),
```
Comment (vmarkovtsev) [inline lookout/style/format/default_config.py]: No "aggregate" here but we are using it above

File: lookout/style/format/tests/bugs/003_classify_vnodes_negative_col/test.py [hunk 5]
```diff
@@ -5,14 +5,14 @@
...
+from lookout.style.format.tests.test_analyzer import get_config
 
 
 class FeaturesTests(unittest.TestCase):
     def setUp(self):
-        config = FormatAnalyzer._load_train_config(get_train_config())
+        config = FormatAnalyzer._load_config(get_config())
         self.extractor = FeatureExtractor(language="javascript",
-                                          **config["javascript"]["feature_extractor"])
+                                          **config["train"]["javascript"]["feature_extractor"])
```
Comment (vmarkovtsev) [inline lookout/style/format/tests/bugs/003_classify_vnodes_negative_col/test.py]: This started to get out of control. How about using [munch](https://github.com/Infinidat/munch)? At least the string constants will go away. We could munchify everything in `_load_config`

File: lookout/style/format/analyzer.py (hunk 1 above)
Comment (zurk) [reply lookout/style/format/analyzer.py]: "common" is also common for all languages. If I place `comment_template` to `analyze` I should get it for each language specifically:
```
with open(self.config["common"]["javascript"]["comment_template"]
```
 That means I should move comment template loading from `__init__` to internals. 
I can do it if you think it is better.

File: lookout/style/format/benchmarks/evaluate_smoke.py (hunk 2 above)
Comment (zurk) [reply lookout/style/format/benchmarks/evaluate_smoke.py]: Yes, it is posible, let's move it out of this PR https://github.com/src-d/style-analyzer/issues/651

File: lookout/style/format/benchmarks/general_report.py (hunk 3 above)
Comment (zurk) [reply lookout/style/format/benchmarks/general_report.py]: because it is language independent. The same as here: https://github.com/src-d/style-analyzer/pull/650#discussion_r

File: lookout/style/format/analyzer.py (hunk 1 above)
Comment (vmarkovtsev) [reply lookout/style/format/analyzer.py]: Not really, "common" is something which is shared between "analyze" and "train". Otherwise, we have "global" inside "analyze" and "train" - btw, we should probably rename it to "common" for clarity.

Comment (vmarkovtsev) [reply lookout/style/format/analyzer.py]: In this case I expect the setting to go into "analyze/global"

File: lookout/style/format/default_config.py (hunk 4 above)
Comment (zurk) [reply lookout/style/format/default_config.py]: it is for quality report analyzer. The default value is defined here: https://github.com/src-d/style-analyzer/pull/650/files#diff-17b9445600e10a2c232356a5f1831885R213

File: lookout/style/format/tests/bugs/003_classify_vnodes_negative_col/test.py (hunk 5 above)
Comment (zurk) [reply lookout/style/format/tests/bugs/003_classify_vnodes_negative_col/test.py]: yes, I like it. but what about munchify everything in next PR?

File: lookout/style/format/analyzer.py (hunk 1 above)
Comment (zurk) [reply lookout/style/format/analyzer.py]: "analyze/global" is global for all languages, yes, but not "common". "analyze/global" is removed to have a structure like: 
```
"analyze": {
    "<lang1>": {},
    "<lang2>": {},
    "<lang3>": {},
}
```
We cannot just keep it because If someone decide to override `comment_template` for one paticular language (and it is possible if we place in "analyze/global") we are able to load only one global template. 

So "global" and "common" are different and I'd like to not have "analyze/common" and "train/common" because it is just make things more complex.

File: lookout/style/format/tests/bugs/003_classify_vnodes_negative_col/test.py (hunk 5 above)
Comment (zurk) [reply lookout/style/format/tests/bugs/003_classify_vnodes_negative_col/test.py]: Also, I think it is still ok to have 3-level nesting because in most cases you deal only with 2 levels and this one is a test.

File: lookout/style/format/analyzer.py (hunk 1 above)
Comment (zurk) [reply lookout/style/format/analyzer.py]: Just to clarify "analyze/global" contains default values for all languages. When we replace unspecified values with defaults we remove "analyze/global"

Comment (vmarkovtsev) [reply lookout/style/format/analyzer.py]: The high level structure must be this: "analyze"- all configuration for `analyze()`. "train" - all configuration for `train()`. "common" must be used for configuration which is shared between train and analyzer, and has nothing to do with languages.

The comment template is not used in `train()` and cannot be put inside "common".

File: lookout/style/format/analyzer.py [hunk 6]
```diff
@@ -233,15 +175,15 @@ def train(cls, ptr: ReferencePointer, config: Mapping[str, Any], data_service: D
...
         for language, files in files_by_language(files).items():
             try:
-                lang_config = config[language]
+                lang_config = config["train"][language]
             except KeyError:
                 _log.warning("language %s is not supported, skipped", language)
                 continue
-            _log.info("effective config for %s:\n%s", language,
-                      pformat(lang_config, width=120, compact=True))
+            _log.debug("effective train config for %s:\n%s", language,
```
Comment (vmarkovtsev) [inline lookout/style/format/analyzer.py]: This *must* stay info. Please do not change unrelated code, otherwise I takes me more time to review and your PRS hang unmerged longer

File: lookout/style/format/analyzer.py [hunk 7]
```diff
@@ -588,29 +532,21 @@ def _group_line_nodes(
...
-                    for lang, lang_config in config.items()}
-        except AttributeError as e:
-            raise ValueError("Config %s can not be merged with default values config: %s: %s" % (
-                config, global_config, e,
-            )) from None
+        effective_config = merge_dicts(cls.default_config, config)
+        for key in ["train", "analyze"]:
+            global_config = effective_config[key].pop("language_defaults")
+            try:
+                for lang in CONFIGURABLE_LANGUAGES:
```
Comment (vmarkovtsev) [inline lookout/style/format/analyzer.py]: Let's avoid yet another pointer to the languages that we support. Try to do it without this variable by enforcing additional constraints.

File: lookout/style/format/analyzer.py [hunk 8]
```diff
@@ -391,8 +334,9 @@ def render_comment_text(self, file_fix: FileFix, fix_index: int) -> str:
             get_change_description, feature_extractor=file_fix.feature_extractor)
         code_lines = file_fix.head_file.content.decode("utf-8", "replace").splitlines()
         line_fix = file_fix.line_fixes[fix_index]
+        config = self.config["analyze"][file_fix.language]
```
Comment (vmarkovtsev) [inline lookout/style/format/analyzer.py]: Put `self.config["analyze"]` to a var

File: lookout/style/format/analyzer.py [hunk 9]
```diff
@@ -189,10 +131,10 @@ def check_training_required(
             data_service.get_data(), old_model.ptr, ptr, contents=True, uast=False))
         base_files_by_lang = files_by_language(c.base for c in changes)
         head_files_by_lang = files_by_language(c.head for c in changes)
-        config = cls._load_train_config(config)
+        config = cls._load_config(config)
         for language, head_files in head_files_by_lang.items():
             try:
-                lang_config = config[language]
+                lang_config = config["train"][language]
```
Comment (vmarkovtsev) [inline lookout/style/format/analyzer.py]: Put `config["train"]` to a var

Comment (zurk) [reply lookout/style/format/analyzer.py]: So do you want something like
```python
train_config = config["train"]
lang_config = train_config[language]
```
Only `lang_config` is used and there is no more use of `train_config`

File: lookout/style/format/analyzer.py (hunk 6 above)
Comment (zurk) [reply lookout/style/format/analyzer.py]: oops, it was unintentionally

File: lookout/style/format/analyzer.py (hunk 9 above)
Comment (vmarkovtsev) [reply lookout/style/format/analyzer.py]: Yep, lgtm

File: lookout/style/format/analyzer.py [hunk 10]
```diff
@@ -588,29 +534,22 @@ def _group_line_nodes(
...
-        config = merge_dicts(cls.defaults_for_train, config)
-        global_config = config.pop("global")
-        try:
-            return {lang: merge_dicts(global_config, lang_config)
-                    for lang, lang_config in config.items()}
-        except AttributeError as e:
-            raise ValueError("Config %s can not be merged with default values config: %s: %s" % (
-                config, global_config, e,
-            )) from None
+        supported_langs = set(get_supported_languages())
```
Comment (vmarkovtsev) [inline lookout/style/format/analyzer.py]: ok, how about making the function return a set then?

File: lookout/style/format/cmdline_tools.py [hunk 11]
```diff
@@ -205,11 +205,8 @@ def add_parser(name, help):
         "--bblfsh",
         help="Babelfish server's address.")
     eval_smoke_parser.add_argument(
-        "--train-config", type=json.loads, default="{}",
-        help="Json config for train step.")
-    eval_smoke_parser.add_argument(
-        "--analyze-config", type=json.loads, default="{}",
-        help=" Json config for analyze step.")
+        "--config", type=json.loads, default="{}",
+        help="Json config for FormatAnalyzer.")
```
Comment (vmarkovtsev) [inline lookout/style/format/cmdline_tools.py]:
```suggestion
        help="JSON config for FormatAnalyzer.")
```

File: lookout/style/format/config.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+def get_supported_languages() -> List[str]:
+    """
+    Return a list of supported languages by style-analyzer.
+    """
+    return [modname for importer, modname, ispkg in pkgutil.iter_modules(langs.__path__) if ispkg]
+
+
+DEFAULT_CONFIG = {
+    "common": {
+        # All common settings for train and analyze should be here
```
Comment (vmarkovtsev) [inline lookout/style/format/config.py]:
```suggestion
        # Shared settings for train and analyze should be here
```

**Comment by vmarkovtsev** [comment]
@zurk Approved. Let's correct those last things and merge
